src/Microsoft.ML.Data/Dirty/ChooseColumnsByIndexTransform.cs

```diff
@@ -31,158 +30,115 @@ public sealed class Arguments
-                        else
-                            throw Contracts.ExceptDecode(fmt, col, _input.ColumnCount);
-                    }
-                }
-                if (drop)
-                {
-                    sources = Enumerable.Range(0, _input.ColumnCount).Except(indexCopy).ToArray();
-                    dropped = indexCopy;
-                }
-                else
+                var schemaBuilder = new SchemaBuilder();
+                for (int i = 0; i < _sources.Length; ++i)
                 {
-                    sources = indexCopy;
-                    dropped = null;
+                    // selectedIndex is an column index of input schema. Note that the input column indexed by _sources[i] in _sourceSchema is sent
+                    // to the i-th column in the output schema.
+                    var selectedIndex = _sources[i];
+
+                    // The dropped/kept columns are determined by user-specified arguments, so we throw if a bad configuration is provided.
+                    string fmt = string.Format("Column index {0} invalid for input with {1} columns", selectedIndex, _sourceSchema.ColumnCount);
+                    Contracts.Check(selectedIndex < _sourceSchema.ColumnCount, fmt);
+
+                    // Copy the selected column into output schema.
+                    var selectedColumn = _sourceSchema[selectedIndex];
+                    schemaBuilder.AddColumn(selectedColumn.Name, selectedColumn.Type, selectedColumn.Metadata);
                 }
-                if (user)
-                    Contracts.CheckUserArg(sources.Length > 0, nameof(Arguments.Index), "Choose columns by index has no output columns");
-                else
-                    Contracts.CheckDecode(sources.Length > 0, "Choose columns by index has no output columns");
-                nameToCol = new Dictionary<string, int>();
-                for (int c = 0; c < sources.Length; ++c)
-                    nameToCol[_input.GetColumnName(sources[c])] = c;
+                return schemaBuilder.GetSchema();
             }
 
-            public Bindings(ModelLoadContext ctx, Schema schemaInput)
+            internal Bindings(ModelLoadContext ctx, Schema sourceSchema)
             {
                 Contracts.AssertValue(ctx);
-                Contracts.AssertValue(schemaInput);
-
-                _input = schemaInput;
+                Contracts.AssertValue(sourceSchema);
 
                 // *** Binary format ***
-                // bool(as byte): whether the indicated source columns are columns to keep, or drop
-                // int: number of source column indices
-                // int[]: source column indices
+                // bool (as byte): operation mode
+                // int[]: selected source column indices
+                _drop = ctx.Reader.ReadBoolByte();
+                _sources = ctx.Reader.ReadIntArray();
```

Comment:
This look suspicious.

So imagine I create transform with params [0,2,3] and drop:false
And we do it for two Schemas, one with 4 columns, one with 5.

result for both of them would be schema with 3 columns [0,2,3].

now let's work with [0,2,3] and drop:true

I pass schema with 4 columns to this transform and save it.
we will save [1] in source.

Now I load this transform and pass schema with 5 columns.
result would be same schema with [1] column.

But if I start on 5 columns schema I will have [1,4] and if I pass that saved transform schema with 4 columns, i'm screwed.


So please, preserve old logic and respect drop parameter.

Reply:
Also would be nice if you throw some test coverage for this transform.

```
  [TestMethod, TestCategory("DataPipeSerialization")]
        public void SavePipeChooseColumnsByIndex()
        {
            TestCore(null, true,
                new[] {
                    "loader=Text{",
                    "  col=Label:U1[0-1]:0",
                    "  col=Features:U2:1-*",
                    "  col=A:U1[1-5]:1",
                    "  col=B:U1[3-8]:2",
                    "}",
                    "xf=ChooseColumnsByIndex{ind=2 ind=0}"
                });

            Done();
        }

        [TestMethod, TestCategory("DataPipeSerialization")]
        public void SavePipeChooseColumnsByIndexDrop()
        {
            TestCore(null, true,
                new[] {
                    "loader=Text{",
                    "  col=Label:U1[0-1]:0",
                    "  col=Features:U2:1-*",
                    "  col=A:U1[1-5]:1",
                    "  col=B:U1[3-8]:2",
                    "}",
                    "xf=ChooseColumnsByIndex{ind=3 ind=0 drop}"
                });

            Done();
        }

```

---
In reply to: [241891989](https://github.com/dotnet/machinelearning/pull/1879#discussion_r241891989) [](ancestors = 241891989)

Reply:
What is `true`? There is no `TestCore` with prototype `TestCore(string, bool, ...)`.

---
In reply to: [241892330](https://github.com/dotnet/machinelearning/pull/1879#discussion_r241892330) [](ancestors = 241892330,241891989)

Reply:
Tests are added and old logic is back.

---
In reply to: [241903270](https://github.com/dotnet/machinelearning/pull/1879#discussion_r241903270) [](ancestors = 241903270,241892330,241891989)